### ISSUE TITLE
can't login to chat on nextcloud

### ISSUE THREAD

**Issue by matthewseal** [jsxc/xmpp-cloud-auth#30]
Hello!

I am trying to set up an external prosody xmpp server setup (on the same machine)
I can't seem to login through nextcloud, although I can login to the xmpp server directly just fine.
The log from the browser seems to tell what is going on,

          Content Security Policy: Ignoring duplicate source 'unsafe-inline'  (unknown)
          Content Security Policy: Couldn’t parse invalid host 'unsafe-inline'  (unknown)
          Content Security Policy: Directive ‘frame-src’ has been deprecated. Please use directive ‘child-src’ instead.  (unknown)
          JQMIGRATE: Migrate is installed, version 1.4.0  core.js:7:542
          State changed to INITIATING  jsxc.js:183:10
          State changed to SUSPEND  jsxc.js:183:10
          [WARN] Unable to create user prefix jsxc.js:183:10
          [WARN] Unable to create user prefix jsxc.js:183:10
          State changed to ESTABLISHING  jsxc.js:183:10
          New connection  jsxc.js:183:10
          CONNECTING: null  jsxc.js:183:10
          close dialog

Any help would be appreciated!

**Comment by MarcelWaldvogel** [comment]
What do the *Prosody* and *xmpp-cloud-auth* log files say?

**Comment by matthewseal** [comment]
The prosody logs have zero errors (since being configured to what the wiki says). prosody.log does show the bosh sessions being created, is there more it should show after that?

xcauth.err is empty
xcauth.log shows: INFO: FAILURE: Could not authenticate user <redacted>: noauth 
a few times but not recently

**Comment by matthewseal** [comment]
I am using apache proxypass through this configuration:

        ProxyPreserveHost On
        ProxyRequests Off
        ProxyPass /http-bind/ http://127.0.0.1:5280/http-bind/
        ProxyPassReverse /http-bind/ http://127.0.0.1:5280/http-bind/

**Comment by MarcelWaldvogel** [comment]
Did `xcauth.log` log any activity at all recently? How are you running it? Which `extauth` plugin/mode?

**Comment by matthewseal** [comment]
I am running it according to the wiki [https://github.com/jsxc/xmpp-cloud-auth/wiki/prosody](https://github.com/jsxc/xmpp-cloud-auth/wiki/prosody) except I changed external_cloud.py to xcauth.py in prosody, as well as using xcauth.conf under /etc/ 

xcauth.log shows some successful logins, a few failed because of "noauth" and then more successful logins. logging in through nextcloud (having to click login without cloud however) does not add to the log

**Comment by MarcelWaldvogel** [comment]
Are you using the experimental time-limited tokens (Nextcloud admin page for JSXC) and are using Nextcloud login names with capitals (or other uppercase characters)?

A fix for that is in [`nextcloud/jsxc.nextcloud@fcc9b7cae`](https://github.com/nextcloud/jsxc.nextcloud/commit/fcc9b7cae7b60a43cd30a5fc1394b8a32ecf1eeb#diff-5e96b74456d4709f63e5c02b1d37a251R91)

**Comment by matthewseal** [comment]
yes I am using logins with upper case characters. I will try and see if I can manually upgrade to the 3.3 version later today. I hope this works and thanks!

**Comment by matthewseal** [comment]
hello! I have upgrade ojsxc to 3.3.0-beta but I still have the same issue :( I don't know what I am doing wrong. I have even tried with ejabberd but with less success.

**Comment by MarcelWaldvogel** [comment]
With the current information, I have to tap in the dark. Let me try:
- Can you log in with a user with lowercase-only account name? (Create a test user if necessary)
- Do you have user IDs with `@` in them?
- What happens when you deactivate the experimental time-limited tokens?
Sorry for the burden

**Comment by matthewseal** [comment]
Hello, I have tried all these cases before, and none fixed the issue.

The main reason I am trying to set this up is for a company I am volunteering at that need secure communication. XMPP by itself actually has everything we need, which is chat and file transfer. The only thing I need to find out know is just encrypting the data while at rest on the server.

**Comment by MarcelWaldvogel** [comment]
Nextcloud does provide you with an option to encrypt data at rest. Some XMPP clients such as [Conversations](https://github.com/siacs/Conversations) do perform (non-standard) encryption while transferring files using XMPP HTTP Upload.

Nevertheless, I would like to know why JSXC is not working for you. Could you provide me with the Nextcloud, Apache (of the proxy), XMPP server, and xcauth log files of your login attempts?

BTW: Did you ever try with a different browser and without plugins? What browser are you currently using? Do you have any other tabs open to the Nextcloud instance? Did you clear the browser cache?

@sualko Do the following messages from the log file at the beginning ring a bell?

```
State changed to SUSPEND  jsxc.js:183:10
[WARN] Unable to create user prefix jsxc.js:183:10
State changed to ESTABLISHING  jsxc.js:183:10
New connection  jsxc.js:183:10
CONNECTING: null  jsxc.js:183:10
```

**Comment by matthewseal** [comment]
I have destroyed the test nextcloud instance, I will recreate it for you and send you the logs
Is a shared link through nextcloud good enough? Makes it easier for me

**Comment by matthewseal** [comment]
So I recreated the instance, same issues as before. Here are the logs:

[apache-access.log.txt](https://github.com/jsxc/xmpp-cloud-auth/files/1206292/apache-access.log.txt)
[apache-error.log.txt](https://github.com/jsxc/xmpp-cloud-auth/files/1206291/apache-error.log.txt)

[prosody.log.txt](https://github.com/jsxc/xmpp-cloud-auth/files/1206290/prosody.log.txt)

**Comment by xekon** [comment]
I am also unable to login to nextcloud, but i did my best to stumble through all possible diagnostic logs and tools I could find.

you can test logging in with debugging if you goto the tool directly:

    cd /opt/xmpp-cloud-auth
    sudo ./xcauth.py -d -A xekon example.com password1234

2018-05-08 23:57:56,019 DEBUG: Token is too short: 14 != 23 (maybe not a token?)

this is my first clue of a major problem.....
so after another look at my token, I noticed that the / character is a part of my token, and so im wondering if that character is maybe splitting my password unintentionally.....

I was hoping I could just regenerate a different Secure API token, but I do not seen an option for that...
Hope some of this info helps.

**Comment by xekon** [comment]
im trying to find where in the database or files that key is located, so far i have not found it.

**Comment by xekon** [comment]
Honestly, with how much conflicting/incorrect information I have found for different config options of Prosody. To just install it with the basic options it seems to work great without much fuss.

but getting this external auth to work with it has been a horrible experience.

I think I will give ejabberd a go, atleast then I can say I tried them both. Initially when I compared many of the listed pros and cons, and compared the programming languages I was more drawn to Prosody, but if ejabberd ends up being easier to setup then I will just use it.
